src/core/exception.d

```diff
@@ -690,15 +785,15 @@ extern (C)
 
     /* Called when an array index is out of bounds
      */
-    void _d_arrayboundsp(immutable(char*) file, uint line)
+    void _d_arrayboundsp(immutable(char*) file, uint line, size_t lower, size_t upper, size_t length)
     {
         import core.stdc.string : strlen;
-        onRangeError(file[0 .. strlen(file)], line);
+        onRangeError(lower, upper, length, file[0 .. strlen(file)], line);
     }
 
-    void _d_arraybounds(string file, uint line)
+    void _d_arraybounds(string file, uint line, size_t lower, size_t upper, size_t length)
```

Comment:
Rather than changing the signatures of the public extern C interface, new functions should be introduced.

Reply:
For reference the DMD patch is already in, hence why I want to get this in. Also as noted by Jacob, we don't guarantee binary compatibility across releases.

Reply:
See `_aaGet`, `_aaGetX`, `_aaGetY` for examples of doing this in the past.  Infact pick any function that ends in X (`_d_arrayliteralX`)

Even `_d_arrayboundsp` was a new name given to replace `_d_arrayboundsm`. :-)

Reply:
But what's the point of doing so and creating more legacy cruft if we don't we don't guarantee binary compatibility across releases anyhow?

Reply:
- Gives compiler vendors time to adjust their codegen and test the new api.
- Doesn't break CI (druntime functions should have been applied first before compiler changes).
- Doesn't break git bisect (Digger) infrastructure.

Reply:
> Gives compiler vendors time to adjust their codegen and test the new api.

Fair point, but wouldn't this apply to new changes?

> Doesn't break CI (druntime functions should have been applied first before compiler changes).

It passed on all CIs.

> Doesn't break git bisect (Digger) infrastructure

Digger is used to build DAutoTest, so that's tested too.

Reply:
> It passed on all CIs.

Question: do you know *why* it passes? Looking at this and the compiler change, I do not.

Reply:
Call parameter stack layout: adding args to the end doesn't affect previous ones. I.e. passing too many will ignore the later ones, whereas passing too few will lead to garbage/corruption.

Reply:
x86 is the only target that pushes arguments in reverse.  All others are left to right.

I have an ARM and Aarch64 server that I can run a test on. Regressions on those targets could warrant a revert... Did the druntime changes end up in a release?

Reply:
Assuming that it's a dmd backend change only though, it's not so critical.  If array bounds checks were being generated by the frontend, then it's another matter.

Reply:
It is a backend thing.